adaptive/runner.py

```diff
@@ -162,15 +170,20 @@ def do_log(self):
         return self.log is not None
 
     def _ask(self, n):
-        points = [
-            p for p in self.to_retry.keys() if p not in self.pending_points.values()
-        ][:n]
-        loss_improvements = len(points) * [float("inf")]
-        if len(points) < n:
-            new_points, new_losses = self.learner.ask(n - len(points))
-            points += new_points
+        pending_ids = self._pending_tasks.values()
+        pids_gen = (pid for pid in self._to_retry.keys() if pid not in pending_ids)
+        pids = list(itertools.islice(pids_gen, n))
```

Comment:
If it is important to keep it like this, please comment about why a generator comprehension is used.

The common idiom would be to use a list comprehension

Reply:
if it is for readability only then a comment is not necessary, and you can just resolve with convo

Reply:
I moved from the list comprehension because not the whole list needs to be created, only until `n`.

Reply:
I added a comment!